### DISCUSSION TITLE
Updating the right status with keyboard layout through shell commands

### DISCUSSION THREAD

**Discussion by McFrappe** [wezterm/wezterm#1632] [Q&A] [answered]
Hello Wez! 

I was wondering if you know if lua should do some sort of async/await in order to fetch information by `io.popen()`? 
What im trying to do is that i have fetched which layout I'm using through: 
`plutil -extract AppleSelectedInputSources xml1 -o - ~/Library/Preferences/com.apple.HIToolbox.plist | xmllint --xpath \"string(//string[2])\" -`
And then wanting to update the right status when im in fullscreen mode so that i know between which layouts im switching between. Although, when actually fetching the layout etc, the right status will not update at all (or at least it seems to hang and dont update?). 

Any idea on this? 

```lua
local function execr(cmd)
    local handle = io.popen(cmd);
    local result = handle:read("*all");
    handle.flush();
    handle.close();
    return result;
end

wezterm.on("update-right-status", function(window)
    local window_dimensions = window:get_dimensions();

    if not window_dimensions.is_full_screen then
        window:set_right_status(wezterm.format({{Text=""},}));
        return;
    else
        local cmdKbdInfo = "plutil -extract AppleSelectedInputSources xml1 -o - ~/Library/Preferences/com.apple.HIToolbox.plist | xmllint --xpath \"string(//string[2])\" -";
        local kbdLayout = execr(cmdKbdInfo);
        local date = wezterm.strftime("%a %b %-d %_I:%M:%S");
        local bat = ""
        for _, b in ipairs(wezterm.battery_info()) do
            bat = string.format("%.0f%%", b.state_of_charge * 100)
        end

        window:set_right_status(wezterm.format({
            {Text= kbdLayout .. "   " .. bat .. "   " .. date},
        }));

        return;
    end
end)
```

**Comment by wez** [comment, marked as answer]
https://wezfurlong.org/wezterm/config/lua/wezterm/run_child_process.html can be used to capture the output from a child process without blocking the gui thread; it internally knows how to do the equivalent of an async/await

**Reply by McFrappe** [reply]
I tried to do this;

```lua
        -- local cmdKbdInfo = "plutil -extract AppleSelectedInputSources xml1 -o - ~/Library/Preferences/com.apple.HIToolbox.plist | xmllint --xpath \"string(//string[2])\" -";
        -- local kbdLayout = execr(cmdKbdInfo);
        local plutil_results = wezterm.run_child_process({"plutil", "-extract", "AppleSelectedInputSources", "xml1", "-o", "-", "~/Library/Preferences/com.apple.HIToolbox.plist"});
        local xmllint_results = wezterm.run_child_process({"xmllint", "--xpath", "string(//string[2])", plutil_results.stdout});
        local date = wezterm.strftime("%a %b %-d %_I:%M:%S");
        local bat = ""
        for _, b in ipairs(wezterm.battery_info()) do
            bat = string.format("%.0f%%", b.state_of_charge * 100)
        end

        window:set_right_status(wezterm.format({
            {Text= xmllint_results.stdout .. "   " .. bat .. "   " .. date},
        }));
```
Did not seem to show up when activating fullscreen. Nothing rendered so it is somehow still blocking the GUI thread as you mentioned?

**Reply by wez** [reply]
do you see errors in the debug overlay? press CTRL-SHIFT-L to show it

**Reply by McFrappe** [reply]
Im doing it in a single child process instead as follows:
```lua
local success, stdout, stderr = wezterm.run_child_process({"plutil", "-extract", "AppleSelectedInputSources", "xml1", "-o", "-", "~/Library/Preferences/com.apple.HIToolbox.plist", "|", "xmllint", "--xpath", "string(//string[2])", "-"});
```

Now when i go into fullscreen it does print out `bat` and `date`, not `stdout`.
Here is the output from the debug overlay:
<img width="1154" alt="image" src="https://user-images.githubusercontent.com/17787601/153573836-76535e40-05d1-436e-ae60-5df6f4e09471.png">
No error revolving around the child-process...

**Reply by McFrappe** [reply]
I tried printing out `stderr` instead to the bar and it seems that the process itself is failing:

![image](https://user-images.githubusercontent.com/17787601/153588558-0b15b959-4278-4b5a-ace4-e9a5d07181f3.png)

Those flags are part of `xmllint` but is it that `wezterm.run_child_process` does not recognise them as flags or?

**Reply by wez** [reply]
`run_child_process` executes the program directly with the arguments you provide.
It looks like you're trying to use `|` to pipe things together: you need to use a shell to interpret the pipes.
You can `run_child_process({"sh", "-c", "plutil -extract ... | xmllint --xpath ... "})` to explicitly run via the shell.

**Reply by McFrappe** [reply]
Ahhhhh FIXED!

**Reply by McFrappe** [reply]
Thank you Wez for putting up with me. I now have it working to some extent.
It is a bit wonky but it displays the layout!
I might have to make so that the right status is updated upon the keybinding for changing layout or some sort, an event trigger.

https://user-images.githubusercontent.com/17787601/153674410-974648b9-2ae0-4b02-848c-5fcf227d134f.mov
